### PR TITLE
Switch to using futures-micro for things it supports. Resolves #6, #7, #8

### LINKED ISSUE #6: join! macro to join more than two futures?
futures-lite implements the `join` function to join two futures. I'd love to have the `join!` macro to join more than two futures, if that'd be a reasonable addition to futures-lite.

### PR DESCRIPTION
**Pull request by jjl**
I ended up moving a few things from here into micro (which now has better docs!).

New features: `or!`, `poll_state`, `zip!`

Everything else is just not duplicating things micro now has and minor tidying.

### CODE REVIEW

**Comment by zserik** [comment]
@bugavec I deprecated these functions (incl. `pending()`, which is also superfluous).

**Comment by taiki-e** [comment]
- This adds a dependency on a lot of unsafe code that doesn't seem to be documented at all.
- Since futures-micro's major version is less than 1, it is basically unacceptable to re-export their types directly. Please wrap them using the new-type pattern.
- This adds additional unneeded `Unpin` requirements to some types of `Unpin` implementations. (breaking changes)

Also, [futures-micro](https://github.com/irrustible/futures-micro)'s readme says something like this:

> Yes, we could get rid of them with pin-project-lite, but it's just hiding the mess and then we couldn't claim zero dependencies.

"just hiding the mess" is *not* correct. pin-project/pin-project-lite guarantees the safety requirements required for pin-projection, at compile time. (See pin-project's [docs](https://docs.rs/pin-project/1/pin_project/attr.pin_project.html#safety) for details)

I've seen a lot of unsound code in a project that made similar claims and using unsafe code before, and I'm worried that futures-micro will be in the same state in the future.

**Comment by jjl** [comment]
This PR is very old, I think  probably the right answer is to close it at this point, but i could bring it up to date if you wanted. Notes:

>     * This adds a dependency on a lot of unsafe code that doesn't seem to be documented at all.

Yes, the code currently assumes the implementer knows what they're doing. Yes, that's suboptimal.

>     * Since futures-micro's major version is less than 1, it is basically unacceptable to re-export their types directly. Please wrap them using the new-type pattern.

This PR predates futures-lite 1.0. Though we will probably release futures-micro 1.0 soon.

>     * This adds additional unneeded `Unpin` requirements to some types of `Unpin` implementations. (breaking changes)

I'll add it to the list of things to look at when I get time.

> Also, [futures-micro](https://github.com/irrustible/futures-micro)'s readme says something like this:
> 
> > Yes, we could get rid of them with pin-project-lite, but it's just hiding the mess and then we couldn't claim zero dependencies.
> 
> "just hiding the mess" is _not_ correct. pin-project/pin-project-lite guarantees the safety requirements required for pin-projection, at compile time. (See pin-project's [docs](https://docs.rs/pin-project/1/pin_project/attr.pin_project.html#safety) for details)
> 
> I've seen a lot of unsound code in a project that made similar claims and using unsafe code before, and I'm worried that futures-micro will be in the same state in the future.

That's interesting to know. At some point I will make a move to pin-project-lite in light of this.

File: src/future.rs [hunk 1]
```diff
@@ -15,32 +15,31 @@
...
-
 use core::fmt;
 #[doc(no_inline)]
 pub use core::future::Future;
-use core::marker::PhantomData;
 use core::pin::Pin;
+use core::task::{Context, Poll};
 
+pub use futures_micro::{
+    or, pending, poll_fn, poll_state, sleep, waker, zip,
```
Comment (zserik) [inline src/future.rs:25]: side note: `pending,poll_state,PollState,Ready` were dropped from `futures-micro`.

**Comment by zserik** [comment]
@jjl about the unpin stuff: I think `pin-project-lite` might have been able to detect the pinning guarantee violation which was present in the past in [`next_poll`](https://docs.rs/futures-micro/0.3.1/futures_micro/fn.next_poll.html)...

**Comment by jjl** [comment]
> That's interesting to know. At some point I will make a move to pin-project-lite in light of this.

This one is now done in main. The rest require a little more thought.

**Comment by notgull** [comment]
I'm going to close this. [https://github.com/irrustible/futures-micro](`futures-micro` hasn't been updated in over a year), and at this point it mostly just re-implements code that's already in here.

Thanks anyways for the PR!

**Pull request closed without merging**